File: src/warm_bridge/tutor.py
"""Strength tutoring — when is it safe to ask this bridge?

Maps strength + mode + confidence to plain-language guidance.
Never invents closeness; uses only fields from the import + scoring.
"""

from __future__ import annotations

from typing import Any

HARD_INTRO_MODES = frozenset({"ask_intro"})
SOFT_MODES = frozenset({"ask_permission", "ask_intel", "peer_forward"})


def strength_advice(
    *,
    strength: str,
    mode: str,
    confidence: str,
    types: list[str],
    bucket: str,
    locale: str = "pt",
) -> dict[str, Any]:
    """Return tutoring payload for UI: level yes | soft | no + headline + bullets."""
    s = (strength or "unknown").lower()
    m = mode or "ask_permission"
    is_direct = bucket == "direct" or "direct" in types

    if is_direct:
        return _pack(
            level="yes",
            can_ask=True,
            locale=locale,
            headline_pt="Você já tem o alvo na rede — pode abordar direto.",
            headline_en="Target is already in your graph — reach out directly.",
            bullets_pt=[
                "Não precisa pedir intro a ninguém.",
                "Use um tom curto e contextual — WhatsApp ou LinkedIn.",
            ],
            bullets_en=[
                "No bridge intro needed.",
                "Keep it short and contextual — WhatsApp or LinkedIn.",
            ],
        )

    if s == "low" or (s == "unknown" and m in HARD_INTRO_MODES and confidence == "low"):
        return _pack(
            level="no",
            can_ask=False,
            locale=locale,
            headline_pt="Ainda não peça intro dura — relação fraca ou fria.",
            headline_en="Don't ask for a hard intro yet — weak or cold tie.",
            bullets_pt=[
                "Prefira pedir mapa (quem decide?) ou permissão leve.",
                "Reaqueça o contato antes de citar o alvo.",
            ],
            bullets_en=[
                "Prefer intel (who decides?) or soft permission.",
                "Re-warm the contact before naming the target.",
            ],
        )

    if m in HARD_INTRO_MODES and s == "high" and confidence in ("high", "medium", "direct"):
        return _pack(
            level="yes",
            can_ask=True,
            locale=locale,
            headline_pt="Sim — confiança alta e caminho claro para intro.",
            headline_en="Yes — high trust and a clear path for an intro.",
            bullets_pt=[
                "Peça uma intro curta com texto pronto para encaminhar.",
                "Dê uma saída fácil se não for o momento.",
            ],
            bullets_en=[
                "Ask for a short intro with a forwardable blurb.",
                "Give them an easy out if timing is off.",
            ],
        )

    if m == "peer_forward" and s in ("high", "medium"):
        return _pack(
            level="soft",
            can_ask=True,
            locale=locale,
            headline_pt="Pode pedir para encaminhar — sem precisar defender a ideia.",
            headline_en="OK to ask them to forward — no need to sell the idea.",
            bullets_pt=[
                "Encaminhar texto é mais leve que pedir intro ativa.",
                "Funciona bem no WhatsApp com relação média/alta.",
            ],
            bullets_en=[
                "Forwarding text is lighter than an active intro.",
                "Works well on WhatsApp with medium/high trust.",
            ],
        )

    if m in SOFT_MODES or s in ("medium", "unknown"):
        return _pack(
            level="soft",
            can_ask=m != "ask_intro",
            locale=locale,
            headline_pt="Pedido leve — mapa, permissão ou encaminhar, não intro dura.",
            headline_en="Soft ask — intel, permission, or forward, not a hard intro.",
            bullets_pt=[
                "Confirme quem decide antes de pedir ponte.",
                "Cite o alvo só se a relação aguentar.",
            ],
            bullets_en=[
                "Confirm who decides before asking for a bridge.",
                "Name the target only if the relationship can handle it.",
            ],
        )

    return _pack(
        level="no",
        can_ask=False,
        locale=locale,
        headline_pt="Espere — fortaleça a relação antes de pedir.",
        headline_en="Hold — strengthen the relationship before asking.",
        bullets_pt=[
            "Adicione notas ou reative o contato.",
            "Tente outra ponte com strength=high.",
        ],
        bullets_en=[
            "Add notes or re-engage the contact.",
            "Try another bridge marked high strength.",
        ],
    )


def attach_tutor(row: dict[str, Any], locale: str = "pt") -> dict[str, Any]:
    """Attach tutor dict to an enriched bridge row (mutates and returns row)."""
    row["tutor"] = strength_advice(
        strength=row.get("strength") or "unknown",
        mode=row.get("mode") or "ask_permission",
        confidence=row.get("confidence") or "low",
        types=row.get("types") or [],
        bucket=row.get("bucket") or "bridge",
        locale=locale,
    )
    return row


def _pack(
    *,
    level: str,
    can_ask: bool,
    locale: str,
    headline_pt: str,
    headline_en: str,
    bullets_pt: list[str],
    bullets_en: list[str],
) -> dict[str, Any]:
    en = locale == "en"
    return {
        "level": level,
        "can_ask": can_ask,
        "headline": headline_en if en else headline_pt,
        "headline_pt": headline_pt,
        "headline_en": headline_en,
        "bullets": bullets_en if en else bullets_pt,
        "bullets_pt": bullets_pt,
        "bullets_en": bullets_en,
    }

File: web/src/tutor.ts
import type { Bridge, TutorAdvice } from "./api";
import type { ReachEvent } from "./outcomes";
import { recentFavorAsk } from "./outcomes";

export function tutorLevelLabel(level: string, locale: string): string {
  const pt: Record<string, string> = { yes: "Sim", soft: "Leve", no: "Ainda não" };
  const en: Record<string, string> = { yes: "Yes", soft: "Soft", no: "Not yet" };
  const map = locale === "en" ? en : pt;
  return map[level] || level;
}

const FAVOR_DAYS = 14;

/** Soft/no if this bridge was asked recently (sent / intro_landed). */
export function applyFavorBank(
  advice: TutorAdvice,
  bridgeId: string,
  outcomes: ReachEvent[],
  locale: string,
): TutorAdvice & { favor_cooldown?: boolean } {
  const hit = recentFavorAsk(bridgeId, outcomes, FAVOR_DAYS);
  if (!hit) return advice;
  const en = locale === "en";
  const when = hit.at.slice(0, 10);
  return {
    ...advice,
    level: advice.level === "no" ? "no" : "soft",
    can_ask: false,
    favor_cooldown: true,
    headline: en
      ? `Cooldown — you already asked ${hit.bridgeName} recently (${when}).`
      : `Já pediu recentemente a ${hit.bridgeName} (${when}).`,
    headline_pt: `Já pediu recentemente a ${hit.bridgeName} (${when}).`,
    headline_en: `Cooldown — you already asked ${hit.bridgeName} recently (${when}).`,
    bullets: en
      ? [
          "Wait before asking the same bridge again.",
          "Pick another path or mark status if this ask already landed.",
        ]
      : [
          "Espere antes de pedir de novo à mesma ponte.",
          "Escolha outro caminho ou atualize o status se o pedido já andou.",
        ],
    bullets_pt: [
      "Espere antes de pedir de novo à mesma ponte.",
      "Escolha outro caminho ou atualize o status se o pedido já andou.",
    ],
    bullets_en: [
      "Wait before asking the same bridge again.",
      "Pick another path or mark status if this ask already landed.",
    ],
  };
}

/** Client fallback when API omits tutor (older server). */
export function fallbackTutor(bridge: Bridge, locale: string): TutorAdvice {
  const s = (bridge.strength || "unknown").toLowerCase();
  const m = bridge.mode || "ask_permission";
  const isDirect = bridge.bucket === "direct";

  if (isDirect) {
    return pack(
      "yes",
      true,
      locale,
      "Você já tem o alvo na rede — pode abordar direto.",
      "Target is already in your graph — reach out directly.",
      ["Não precisa pedir intro a ninguém.", "Tom curto no WhatsApp ou LinkedIn."],
      ["No bridge intro needed.", "Keep it short on WhatsApp or LinkedIn."],
    );
  }
  if (s === "low") {
    return pack(
      "no",
      false,
      locale,
      "Ainda não peça intro dura — relação fraca ou fria.",
      "Don't ask for a hard intro yet — weak or cold tie.",
      ["Prefira mapa ou permissão leve.", "Reaqueça antes de citar o alvo."],
      ["Prefer intel or soft permission.", "Re-warm before naming the target."],
    );
  }
  if (m === "ask_intro" && s === "high") {
    return pack(
      "yes",
      true,
      locale,
      "Sim — confiança alta e caminho claro para intro.",
      "Yes — high trust and a clear path for an intro.",
      ["Peça intro curta com texto pronto.", "Dê saída fácil se não for o momento."],
      ["Ask for a short intro with forwardable text.", "Give an easy out."],
    );
  }
  return pack(
    "soft",
    m !== "ask_intro",
    locale,
    "Pedido leve — mapa, permissão ou encaminhar.",
    "Soft ask — intel, permission, or forward.",
    ["Confirme quem decide antes de pedir ponte.", "Cite o alvo só se a relação aguentar."],
    ["Confirm who decides first.", "Name the target only if the tie can handle it."],
  );
}

export function bridgeTutor(
  bridge: Bridge,
  locale: string,
  outcomes: ReachEvent[] = [],
): TutorAdvice & { favor_cooldown?: boolean } {
  const base = bridge.tutor ?? fallbackTutor(bridge, locale);
  return applyFavorBank(base, bridge.contact_id, outcomes, locale);
}

function pack(
  level: string,
  canAsk: boolean,
  locale: string,
  headlinePt: string,
  headlineEn: string,
  bulletsPt: string[],
  bulletsEn: string[],
): TutorAdvice {
  const en = locale === "en";
  return {
    level,
    can_ask: canAsk,
    headline: en ? headlineEn : headlinePt,
    headline_pt: headlinePt,
    headline_en: headlineEn,
    bullets: en ? bulletsEn : bulletsPt,
    bullets_pt: bulletsPt,
    bullets_en: bulletsEn,
  };
}
